arangod/StorageEngine/TransactionState.cpp

```diff
@@ -528,6 +529,7 @@ void TransactionState::chooseReplicas(
 }
 
 ServerID TransactionState::whichReplica(ShardID const& shard) {
+  MUTEX_LOCKER(guard, _replicaMutex);
```

Comment:
This is not safe now, because `whichReplica` acquires the `_replicaMutex`, but it can call `chooseReplicas` under the mutex, which will acquire it again and deadlock.